### PR TITLE
Unattend refinements

### PR DESCRIPTION
**Pull request by flying-sausages**
## Description

Various issues and potential problems with the unattended installer

## Fixes issues: 
- Issue #nr...
- Un-tracked issue...

## Proposed Changes:
- Start using `sudo` group for master instead of sudoers file
   - accompanying updater
- Escape all passwords being passed around in `box`
- Allow cracklib skip with `--skip-cracklib`
- Install `uuid-runtime` early as apt functions depend on it
- Fixed `env` file mechanism

## Change Categories

- Bug fix 
- Breaking change 

## Checklist

- [ ] Docs have been made OR are not necessary
    - PR link: 
- [ ] Changes to panel have been made OR are not necessary
    - PR link: 
- [ ] Code is formatted [(See more)](https://github.com/swizzin/swizzin/blob/master/CONTRIBUTING.md#editor-plugins-and-tooling)
- [ ] Shellcheck isn't screaming [(See more)](https://github.com/swizzin/swizzin/blob/master/CONTRIBUTING.md#editor-plugins-and-tooling)
- [ ] Prints to terminal are handled [(See more)](https://github.com/swizzin/swizzin/blob/master/CONTRIBUTING.md#printing-into-the-terminal)
- [ ] I have commented my code, particularly in hard-to-understand areas

### Architectures

|   			| `amd64` 	| `armhf` 	| `arm64` 	| Unspecified 	|
|--------		|-------- 	|-------- 	|-------- 	|----------		|
| Focal 		|	✅ 		|		 	|	✅ 		|				|
| Bionic		|			|			|			|				|
| Buster		|	✅		|			|			|				|
| Stretch		|			|			|			|				|
| Raspbian  	|	⚫️		|			|	⚫️		|	⚫️			|

### ✅❎ Passed
- env installation
- skipcracklib
- sudo works after install
- passwords change in torrent clients properly
- no bad prints
- uuid-runtime installed from the getgo

### CODE REVIEW

**Review by liaralabs — DISMISSED** [review]

File: scripts/update/sudoers.sh [hunk 1]
```diff
@@ -8,6 +8,7 @@ if [[ ! -f /etc/sudoers.d/env_keep ]] && [[ $distribution = "Ubuntu" ]]; then
     echo 'Defaults  env_keep -="HOME"' > /etc/sudoers.d/env_keep
 fi
 
+# TODO this should probably skip the master?
```
Comment (flying-sausages) [inline scripts/update/sudoers.sh:11]: Anyy comments on this @liaralabs ?

File: scripts/update/sudoers.sh [hunk 2]
```diff
@@ -48,3 +49,8 @@ www-data     ALL = (ALL) NOPASSWD: CLEANMEM, SYSCMNDS, GENERALCMNDS
 SUD
     fi
 fi
+
+if ! groups "$master" | grep -q sudo; then
+    usermod -a -G sudo "$master"
+    # TODO remove the sudoers file? what about if this runs again, and the master is present in the userlist?
```
Comment (flying-sausages) [inline scripts/update/sudoers.sh:55]: Should we change this?

File: setup.sh [hunk 3]
```diff
@@ -144,18 +147,18 @@ function _option_parse() {
...
-                echo_info "$i added to install queue 1"
+                echo_log_only "$i added to install queue 1"
                 touch /tmp/."$i".lock
             else
                 echo "$i" >> /root/results2
-                echo_warn "$i added to install queue 2"
+                echo_log_only "$i added to install queue 2"
             fi
         done
     fi
```
Comment (flying-sausages) [inline setup.sh:164]: Is this logic good enough for the package sorting?

File: scripts/box [hunk 4]
```diff
@@ -227,42 +228,44 @@ function _adduser() {
...
     fi
 
-    if [[ $(cut -d: -f1 < /root/.master.info 2> /dev/null) = "$user" ]] && [[ -z $SETUP_USER ]]; then
-        echo_error "User already exists within swizzin"
-        exit 1
+    if [[ -z $SETUP_USER ]]; then #If this is not coming from setup.sh...
+        if [[ "$(cut -d: -f1 < /root/.master.info 2> /dev/null)" = "$user" ]]; then
+            echo_error "User already exists within swizzin"
+            exit 1
+        fi
```
Comment (flying-sausages) [inline scripts/box:235]: This bit still needs to get tested

File: scripts/box [hunk 5]
```diff
@@ -227,42 +228,44 @@ function _adduser() {
...
     fi
 
     if id -u "${user}" > /dev/null 2>&1; then
-        echo_info "The user ${user} already appears to be present on this machine; however, it does not appear to be configured as a swizzin user"
-        ask "Continue setting up user?" N || exit 1
+        echo_info "The user ${user} already appears to be present on this machine; however, it does not appear to be configured as a swizzin user.\nThe user will be added to swizzin."
+        ask "Continue setting up user?" Y || exit 1
```
Comment (flying-sausages) [inline scripts/box:240]: Should there even be confirmation logic here?

File: scripts/box [hunk 6]
```diff
@@ -295,12 +298,11 @@ function _adduser() {
 
     #shellcheck source=sources/functions/utils
     . /etc/swizzin/sources/functions/utils
+
     if [[ -n $SETUP_USER ]]; then
-        if ! grep -q -w "${user}" /etc/sudoers.d/swizzin 2> /dev/null; then # Don't add user to sudoers if file already exists in case of botched install
-            echo "${user}	ALL=(ALL:ALL) ALL" >> /etc/sudoers.d/swizzin
-        fi
-        echo "$user:$pass" > /root/.master.info
+        usermod -a -G sudo "$user"
```
Comment (flying-sausages) [inline scripts/box:303]: This should get tested thoroughly on VMs where permissions aren't fucked

File: scripts/box [hunk 7]
```diff
@@ -389,25 +391,25 @@ function _deluser() {
...
 
     master=$(cut -d: -f1 < /root/.master.info)
-    if [[ $user = $master ]]; then
+    if [[ $user = "$master" ]]; then
         echo "$user:$pass" > /root/.master.info
     fi
     echo "$user:$pass" > /root/${user}.info
     chpasswd <<< "${user}:${pass}" >> $log 2>&1
-    htpasswd -b /etc/htpasswd $user $pass >> $log 2>&1
+    htpasswd -b /etc/htpasswd $user "$pass" >> $log 2>&1
```
Comment (flying-sausages) [inline scripts/box:405]: Is this a dangerous change?

File: scripts/box [hunk 8]
```diff
@@ -433,7 +435,7 @@ function _chpasswd() {
     fi
     if [[ -f /install/.qbittorrent.lock ]]; then
         . /etc/swizzin/sources/functions/qbittorrent
-        qbittorrent_chpasswd ${user} ${pass}
+        qbittorrent_chpasswd "${user}" "${pass}"
```
Comment (flying-sausages) [inline scripts/box:438]: This will work I'm assuming?

File: scripts/box [hunk 9]
```diff
@@ -389,25 +391,25 @@ function _deluser() {
...
+    htpasswd -b /etc/htpasswd $user "$pass" >> $log 2>&1
     mkdir -p /etc/htpasswd.d/
-    htpasswd -b -c /etc/htpasswd.d/htpasswd.${user} $user $pass >> $log 2>&1
+    htpasswd -b -c /etc/htpasswd.d/htpasswd.${user} $user "$pass" >> $log 2>&1
     if [[ -e /install/.deluge.lock ]]; then
         systemctl stop deluged@${user} >> $log 2>&1
         systemctl stop deluge-web@${user} >> $log 2>&1
         DWSALT=$(tr -dc 'a-zA-Z0-9' < /dev/urandom | fold -w 32 | head -n 1)
-        DWP=$(python2 /usr/local/bin/swizzin/deluge.Userpass.py ${pass} ${DWSALT})
+        DWP=$(python2 /usr/local/bin/swizzin/deluge.Userpass.py "${pass}" "${DWSALT}")
```
Comment (flying-sausages) [inline scripts/box:412]: This should be tested

File: scripts/update/sudoers.sh (hunk 2 above)
Comment (flying-sausages) [reply scripts/update/sudoers.sh:55]: "this" was meant to be the entire updater script

File: scripts/box (hunk 4 above)
Comment (flying-sausages) [reply scripts/box:235]: works fine

File: scripts/box (hunk 7 above)
Comment (flying-sausages) [reply scripts/box:405]: nope

File: scripts/box (hunk 8 above)
Comment (flying-sausages) [reply scripts/box:438]: tested anbd verified, just needs to aactually restart the service again afterwards though

File: scripts/box (hunk 9 above)
Comment (flying-sausages) [reply scripts/box:412]: good

**Comment by flying-sausages** [comment]
Ready for review

File: scripts/box (hunk 5 above)
Comment (liaralabs) [reply scripts/box:240]: I think we should keep it for now as I think users should be aware that there may be quirks if they have already setup a user

File: scripts/update/sudoers.sh (hunk 1 above)
Comment (liaralabs) [reply scripts/update/sudoers.sh:11]: It already does skip master?

File: scripts/update/sudoers.sh (hunk 2 above)
Comment (liaralabs) [reply scripts/update/sudoers.sh:55]: May as well if we are moving setup that direction

File: sources/functions/users [hunk 10]
```diff
@@ -108,22 +108,27 @@ function password_check() {
...
-            read -n 1 -r
+            echo_progress_done "Password passed"
         fi
-        pass=
-        unset pass
-        password_check "$1"
-    else
-        echo_progress_done "Password passed"
+        SKIPCRACKLIB=""
+        unset SKIPCRACKLIB
```
Comment (liaralabs) [inline sources/functions/users:131]: Have you tested this? In my experience, once cracklib is installed passwords will start failing to set if they don't meet cracklib requirements. Meaning cracklib cannot be installed at all, not just bypassed

Comment (liaralabs) [reply sources/functions/users:131]: Hmmm, seems to work with a generic "password" password under Buster

**Pull request merged**